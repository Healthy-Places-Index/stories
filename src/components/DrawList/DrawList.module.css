.title {
  color: rgb(39, 39, 39);
  font-size: 15px;
  padding-top: 5px;
} 

.icon{
  height: 16px;
  width: 16px;
  margin: 8px 4px 0 0;
}

.input{
  display: grid;
  grid-template-columns: 20px 1fr 80px;
  column-gap: 5px;
}
