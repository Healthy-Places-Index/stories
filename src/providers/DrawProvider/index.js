import React, { useContext, createContext, useReducer, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { useQuery, useMutation, gql } from '@apollo/client';
import { last } from 'lodash';
import { DrawPolygonMode, DrawLineStringMode, EditingMode, DrawPointMode } from 'react-map-gl-draw';

const StateContext = createContext();
const DispatchContext = createContext();

const MODES = [
  { id: 'drawPolyline', text: 'Draw Polyline', Handler: DrawLineStringMode },
  { id: 'drawPolygon', text: 'Draw Polygon', Handler: DrawPolygonMode },
  { id: 'editing', text: 'Edit Feature', Handler: EditingMode },
  { id: 'drawPoint', text: 'Edit Feature', Handler: DrawPointMode },
];

const INITIAL_STATE = {
  mode: null,
  features: [],
  ids: [],
  selectedFeatureIndex: '',
  clickRadius: 12,
  slide: null,
};

const GET_ANNOTATIONS = gql`
  query GetAnnotations($slide: ID!) {
    Slide(where: { id: $slide }) {
      id
      annotations {
        id
        title
        feature
        index
      }
    }
  }
`;

const CREATE_ANNOTATION = gql`
  mutation CreateAnnotation(
    $title: String
    $index: Int
    $feature: String
    $slide: SlideRelateToOneInput
  ) {
    createAnnotation(data: { title: $title, index: $index, feature: $feature, slide: $slide }) {
      id
    }
  }
`;

const UPDATE_ANNOTATION_FEATURE = gql`
  mutation UpdateAnnotation($id: ID!, $feature: String) {
    updateAnnotation(id: $id, data: { feature: $feature }) {
      id
    }
  }
`;

function reducer(state, [type, payload]) {
  // console.log(type, payload);
  switch (type) {
    case 'SET_MODE': {
      const modeId = payload;
      const mode = MODES.find(m => m.id === modeId);
      const modeHandler = mode ? new mode.Handler() : null;
      return { ...state, mode: modeHandler };
    }
    case 'SET_FEATURES': {
      return { ...state, features: payload };
    }
    case 'SET_IDS': {
      return { ...state, ids: payload };
    }
    case 'SET_SELECTED_FEATURE_INDEX': {
      return { ...state, selectedFeatureIndex: payload };
    }
    case 'DEL_FEATURE': {
      return {
        ...state,
        features: state.features.filter((feature, index) => payload !== index),
      };
    }
    case 'SLIDE': {
      return {
        ...state,
        slide: payload,
      };
    }
    default: {
      throw new Error(`Unhandled action type: ${type}`);
    }
  }
}

function DrawProvider({ children }) {
  const [state, dispatch] = useReducer(reducer, INITIAL_STATE);
  const [createAnnotation] = useMutation(CREATE_ANNOTATION);
  const [updateAnnotation] = useMutation(UPDATE_ANNOTATION_FEATURE);
  const { data: annotations } = useQuery(GET_ANNOTATIONS, { variables: { slide: state.slide } });

  const updateTimer = useRef();

  function onSelect(e) {
    const { selectedFeatureIndex } = e;
    dispatch(['SET_SELECTED_FEATURE_INDEX', selectedFeatureIndex]);
  }

  function onUpdate(e) {
    const { data, editType } = e;

    dispatch(['SET_FEATURES', data]);

    switch (editType) {
      case 'addFeature': {
        dispatch(['SET_MODE', 'editing']);
        const newFeature = last(data);
        const index = data.length - 1;
        createAnnotation({
          variables: {
            title: `Feature ${index}`,
            feature: JSON.stringify(newFeature),
            index,
            slide: {
              connect: {
                id: state.slide,
              },
            },
          },
          refetchQueries: [{ query: GET_ANNOTATIONS, variables: { slide: state.slide } }],
        });
        break;
      }
      case 'movePosition': {
        const { ids, selectedFeatureIndex } = state;
        const id = ids[selectedFeatureIndex];
        const feature = JSON.stringify(data[selectedFeatureIndex]);
        clearTimeout(updateTimer.current);
        updateTimer.current = setTimeout(
          () =>
            updateAnnotation({
              variables: { id, feature },
              refetchQueries: [{ query: GET_ANNOTATIONS, variables: { slide: state.slide } }],
            }),
          500
        );
        break;
      }
      default:
        break;
    }
  }

  useEffect(() => {
    if (annotations) {
      dispatch(['SET_FEATURES', annotations.Slide.annotations.map(a => JSON.parse(a.feature))]);
      dispatch(['SET_IDS', annotations.Slide.annotations.map(a => a.id)]);
    }
  }, [annotations]);

  return (
    <StateContext.Provider value={{ ...state, onUpdate, onSelect }}>
      <DispatchContext.Provider value={dispatch}>{children}</DispatchContext.Provider>
    </StateContext.Provider>
  );
}

function useDrawState() {
  const context = useContext(StateContext);

  if (context === undefined) {
    throw new Error('useDrawState must be used within a DrawProvider');
  }

  return context;
}

function useDrawDispatch() {
  const context = useContext(DispatchContext);

  if (context === undefined) {
    throw new Error('useDrawDispatch must be used within a DrawProvider');
  }

  return context;
}

function useDraw() {
  return [useDrawState(), useDrawDispatch()];
}

DrawProvider.propTypes = {
  children: PropTypes.string.isRequired,
};

export { DrawProvider, useDrawState, useDrawDispatch, useDraw };
