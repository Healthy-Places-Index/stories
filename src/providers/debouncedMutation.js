const debouncedMutation = ({ slide, timerRef, mutation, values, typename }) => {
  // eslint-disable-next-line no-underscore-dangle
  const __typename = typename || 'Slide';
  clearTimeout(timerRef.current);
  return setTimeout(
    () =>
      mutation({
        variables: {
          slide,
          ...values,
        },
        optimisticResponse: {
          __typename: 'Mutation',
          updateSlide: {
            __typename,
            id: slide,
            ...values,
          },
        },
      }),
    200
  );
};

export default debouncedMutation;
