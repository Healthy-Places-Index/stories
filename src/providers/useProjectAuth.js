import axios from 'axios';

export default async ({ req, project }) => {
  if (!req.user) return 401;
  const {
    data: {
      data: {
        Project: { user },
      },
    },
  } = await axios.post(`${req.protocol}://${req.get('Host')}/admin/api`, {
    query: `query GetProjectUser($project: ID!) {
        Project(where: { id: $project }) {
          user {
            id
          }
        }
      }
    `,
    variables: {
      project,
    },
  });

  let statusCode = null;
  if (user.id !== req.user.id) statusCode = 403;

  return statusCode;
};
