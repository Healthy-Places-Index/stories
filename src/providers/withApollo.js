/* eslint-disable react/display-name */
/* eslint-disable react/prop-types */
import React from 'react';
import withApollo from 'next-with-apollo';
import { ApolloProvider, ApolloClient, InMemoryCache } from '@apollo/client';

export default withApollo(
  ({ headers }) => {
    return new ApolloClient({
      uri: '/admin/api',
      cache: new InMemoryCache(),
      request: operation => {
        operation.setContext({
          fetchOptions: {
            credentials: 'include',
          },
          headers: { cookie: headers && headers.cookie },
        });
      },
    });
  },
  {
    render: ({ Page, props }) => {
      return (
        <ApolloProvider client={props.apollo}>
          <Page {...props} />
        </ApolloProvider>
      );
    },
  }
);
