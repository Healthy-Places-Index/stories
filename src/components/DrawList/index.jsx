/* eslint-disable react/no-array-index-key */
/* eslint-disable jsx-a11y/label-has-associated-control */
import React from 'react';
import { Button, List, Segment, Form } from 'semantic-ui-react';

import { useDraw } from '../../providers/DrawProvider';
import styles from './DrawList.module.css';

export default function Toolbar() {
  const [{ features }, dispatch] = useDraw();

  if (!features || features.length === 0) return null;
  return (
    <Form.Field>
      <label>Features</label>
      <Segment>
        <List divided verticalAlign="middle">
          {features.map((feature, index) => (
            <List.Item key={index}>
              <List.Content floated="right">
                <Button size="tiny" onClick={() => dispatch(['SET_SELECTED_FEATURE_INDEX', index])}>
                  Edit
                </Button>
                <Button size="tiny" onClick={() => dispatch(['DEL_FEATURE', index])}>
                  Delete
                </Button>
              </List.Content>
              <List.Content verticalAlign="middle" className={styles.title}>
                Feature {index + 1}
              </List.Content>
            </List.Item>
          ))}
        </List>
      </Segment>
    </Form.Field>
  );
}
