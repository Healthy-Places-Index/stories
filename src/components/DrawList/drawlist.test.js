import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { MockedProvider } from '@apollo/client/testing';
import { DrawProvider } from '../../providers/DrawProvider';

import DrawList, { GET_ANNOTATIONS } from './index';

const mocks = [
  {
    request: {
      query: GET_ANNOTATIONS,
      variables: {
        slide: '1',
      },
    },
    result: {
      data: {
        Slide: {
          id: '1',
          annotations: [
            {
              id: '2',
              feature: `{"type":"Feature","properties":{"title":"Title"},"geometry":{"type":"Point","coordinates":[-43.19,-22.92]},"id":"2"}`,
            },
          ],
        },
      },
    },
  },
];

jest.mock('semantic-ui-react', () => {
  const Button = props => <div {...props} className="button" />;
  const Segment = props => <div {...props} className="segment" />;
  const Field = props => <div {...props} className="field" />;
  const Input = props => <input {...props} />;
  const List = props => <div {...props} className="list" />;
  const Item = props => <div {...props} className="list-item" />;
  const Content = props => <div {...props} className="list-content" />;
  List.Item = Item;
  List.Content = Content;
  return { Button, Segment, Form: { Field }, List, Input };
});

describe('Draw List', () => {
  it('matches snapshot', async () => {
    const component = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <DrawProvider>
          <DrawList slide="1" />
        </DrawProvider>
      </MockedProvider>
    );

    await act(() => new Promise(resolve => setTimeout(resolve, 0)));
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
