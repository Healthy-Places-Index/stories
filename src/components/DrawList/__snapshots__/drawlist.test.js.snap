// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Draw List matches snapshot 1`] = `
<div
  className="field"
>
  <label>
    Features
  </label>
  <div
    className="segment"
  >
    <div
      className="list"
      divided={true}
      verticalAlign="middle"
    >
      <div
        className="list-item"
      >
        <div
          className="list-content"
        >
          <div
            className="input"
          >
            <img
              alt="Point"
              className="icon"
              src="/img/Point.svg"
            />
            <input
              disabled={true}
              onChange={[Function]}
              size="mini"
              value="Title"
            />
            <div>
              <div
                className="button"
                disabled={true}
                icon="edit outline"
                onClick={[Function]}
                size="tiny"
              />
              <div
                className="button"
                disabled={true}
                icon="trash alternate"
                onClick={[Function]}
                size="tiny"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
